alibi_detect/od/knn.py

```diff
@@ -0,0 +1,130 @@
+from typing import Callable, Union, Optional, List, Dict, Any
+
+from typing_extensions import Literal
+import numpy as np
+
+from alibi_detect.od.base import OutlierDetector, TransformProtocol, FittedTransformProtocol
+from alibi_detect.od.backend import KNNTorch, AccumulatorTorch
+from alibi_detect.utils.frameworks import BackendValidator
+
+
+backends = {
+    'pytorch': (KNNTorch, AccumulatorTorch)
+}
+
+
+class KNN(OutlierDetector):
+    def __init__(
+        self,
+        k: Union[int, np.ndarray],
+        kernel: Optional[Callable] = None,
+        normalizer: Optional[Union[TransformProtocol, FittedTransformProtocol]] = None,
+        aggregator: Optional[TransformProtocol] = None,
+        backend: Literal['pytorch'] = 'pytorch'
+    ) -> None:
+        """
+        k-Nearest Neighbours (kNN) outlier detector.
+
+        Parameters
+        ----------
+        k
+            Number of nearest neighbours to use for outlier detection. If an array is passed, an aggregator is required
+            to aggregate the scores.
+        kernel
+            Kernel function to use for outlier detection. If None, `torch.cdist` is used.
+        normalizer
+            Normalizer to use for outlier detection. If None, no normalisation is applied.
+        aggregator
+            Aggregator to use for outlier detection. If None, no aggregation is applied. If an array is passed for `k`,
+            then an aggregator is required.
+        backend
+            Backend used for outlier detection. Defaults to `'pytorch'`. Options are `'pytorch'`.
+
+        Raises
+        ------
+        ValueError
+            If `k` is an array and `aggregator` is None.
+        NotImplementedError
+            If choice of `backend` is not implemented.
+        """
+        super().__init__()
+
+        backend_str: str = backend.lower()
+        BackendValidator(
+            backend_options={'pytorch': ['pytorch']},
+            construct_name=self.__class__.__name__
+        ).verify_backend(backend_str)
+
+        if isinstance(k, (list, np.ndarray)) and aggregator is None:
+            raise ValueError((f'k={k} is type {type(k)} but aggregator is {aggregator}, you must '
+                              'specify at least an aggregator if you want to use the knn detector '
+                              'ensemble like this.'))
+
```

Comment:
I've changed the KNN signature so that the aggregator and normalizer both accept either instance of aggregator and normalizer objects or Literals. The defaults are then set to `'ShiftAndScaleNormalizerTorch'` and `'AverageAggregatorTorch'`.  The normalizer can still be set to None in the case where the user doesn't want that functionality for some reason. This was done to ensure that `None` behaves as expected. i.e. to avoid the case where a user selects `normalizer=None` but it's then reset to a default argument. This also seems to be the typical solution, i.e. in tensorflow for instance.